gr-filter/lib/gen_interpolator_taps/objective_fct.c

```diff
@@ -89,10 +91,19 @@ integrand (double omega)
 double
 c_fcn (double *x, int n)
 {
+  gsl_function F;
+  double result, error;
+
+  F.function = integrand;
+  F.params = NULL;
+
   assert ((n & 1) == 0);	/* assert n is even */
   global_n = n;
   global_h = x;
-  return qsimp (integrand, -2 * M_PI * global_B, 2 * M_PI * global_B);
+  gsl_integration_qag(&F, -2 * M_PI * global_B, 2 * M_PI * global_B,
+                      0.0, 1e-12, 1000, GSL_INTEG_GAUSS61,
```

Comment:
This is the only line I'm not clear about: How did you choose those parameters? To be clear, I totally trust your judgment,but from the diff it's not obvious how you reached this. It might improve readability if you declare your epsabs, epsrel, limit as local `const`s. Then again, this is not critical.

Reply:
Well this is an offline utility that is run once every 15 years, so being time efficient does matter.  Keeping that in mind:
- GSL_INTEG_GAUSS61 was the option for best accuracy that this function call had to offer
- epsabs was set to 0.0, so that only epsrel would matter (keeping it simple).
( https://www.gnu.org/software/gsl/manual/html_node/Numerical-Integration-Introduction.html#Numerical-Integration-Introduction )
- epsrel was set to "a really freaking small number" for relative error
- limit was set to "more iterations than we should ever need" to converge
(which wasn't quite right as it specifies the maximum number of subintervals, but the function is so smooth, the number of subintervals shouldn't be large either.)

TL;DR Numerical accuracy mattered; speed didn't.  Let's throw in some extreme numbers.  Oh look, the numbers came out almost the same, and the GSL optimized routines worked faster than the simpson's rule code too. Done. :)

Reply:
I just checked the estimated errors reported by the integration function after it has run.  The largest is 8.136707e-14; the smallest is 1.783098e-24.

I also can't make it better by pushing epsrel smaller (say to 1e-14), since the integration routine will abort() if it can't make the tolerance.

So hey, 1e-12 was a nice lucky guess. :)

Reply:
You have me convinced.